rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+
+One major reason we want to "refresh" the Short URL Service and the URL
+Generator Service is their architecture.
+
+Currently, the Short URL Service is implemented on
+top of `url` type saved object on the server-side. However, it only exposes the
+HTTP endpoints, it does not expose any API on the server for the server-side
+plugins to consume; on the client-side there is no plugin API either, developers
+need to manually execute HTTP requests. 
+
+The URL Generator Service is only available on the client-side, there is no way
+to use it on the server-side, yet we already have use cases (for example ML
+team) where a server-side plugin wants to use a URL generator.
+
+![Current Short URL Service and URL Generator Service architecture](../images/url_service/old_architecture.png)
+
+The current architecture does not allow both service to be conveniently used,
+also as they are implemented in different locations, they are disjointed&mdash;
+we cannot create a short URL using an URL generator.
+
+
+# Detailed design
+
+
+## High level architecture
+
+Below diagram shows the proposed architecture of the URL Service.
+
+![URL Service architecture](../images/url_service/new_architecture.png)
+
+
+## Plugin contracts
+
+The is to provide developers the same experience on the server and browser.
+
+Below are preliminary interfaces of the new URL Service. `IUrlService` will be
+a shared interface defined in `/common` folder shared across server and browser.
+This will allow us to provide users a common API interface on the server and
+browser, wherever they choose to use the URL Service:
+
+```ts
+/**
+ * Common URL Service client interface for the server-side and the client-side.
+ */
+interface IUrlService {
+  generators: IUrlGeneratorClient;
+  shortUrls: IShortUrlClient;
+}
+```
+
+The URL generator client will basically be a registry of URL generators:
+
+```ts
+/**
+ * URL generator registry.
+ */
+interface IUrlGeneratorClient {
+  register<S>(urlGenerator: UrlGeneratorDefinition<S>): UrlGenerator;
+  get<S>(id: string): UrlGenerator<S>;
+}
+
+interface UrlGeneratorDefinition<State> {
+  id: string;
+  createUrl: (state: State) => Promise<string>;
+  isDeprecated?: boolean;
+  migrate?: (state: State) => Promise<{ state: unknown; id: string }>;
+}
+
+interface UrlGenerator<State> {
+  isDeprecated: boolean;
+  createUrl(state: State): Promise<URL>;
+  migrate(): Promise<never | { state: unknown; id: string }>;
+}
+```
+
+The short URL client `IShortUrlClient` which will be the same on the server and
+browser. However, the server and browser might add extra utility methods for
+convenience.
+
+
+```ts
+/**
+ * CRUD-like API for short URLs.
+ */
+interface IShortUrlClient {
+  /**
+   * Create a new short URL from URL generator.
+   *
+   * @param urlGeneratorId ID of the URL generator.
+   * @param state State object to provide to the URL generator.
+   * @param slug Optionally, provide a custom slug (ID) of the short URL.
+   */
+  create(
```

Comment:
this should be discouraged (or should maybe not exist at all) as it promotes introducing non-explicit dependencies. Every plugin providing a generator should expose it on its public contract and every plugin depending on it should use/import it from there. The only exception might be general code that is not aware with what generators it might be working.

Take a look at embeddables for example. Lens embeddable is exposed on lens plugin and should be imported from there whenever you plan to embed lens. Dashboard however doens't know with what kind of embeddables it works with and is thus directly using the registry.

Without this in discover, where we know we rely on visualize plugin for the top chart we would not have it listed in required dependencies (as we would be relying on embeddable). user could disable the visualize plugin and with it break discover (which would expect that visualize embeddable is in the registry, but suddenly isnt

Reply:
cc @stacey-gammon we still follow this mindset right ?

Reply:
@ppisljar, without this or a similar API, how would you link between, let's say, Discover <-> Dashboard ? I think cyclic deps are not allowed.

I think we need API like this. Just types should be explicit that this could throw or return null if URL generator is missing.

Reply:
good point, but introducing api like this we would still be introducing cyclic dependencies just in non-explicit way.

Reply:
I think creating links between two apps is a valid use-case. If we can't express this relation via the core's plugin system (not sure optional plugins can be cyclic), then we will need this registry-like API

Reply:
I think you both have good points, that's why in the current and proposed API you can get hold of a URL generator in two ways: (1) immediately when you create it; (2) through a registry:

![image](https://user-images.githubusercontent.com/9773803/113265412-662ed780-92d4-11eb-9692-c101a2bcdf35.png)

1. In the first use case a plugin would itself expose its URL generator from its plugin contract, which Peter suggested.
2. In the second use case, if there are cyclical dependencies, or for whatever other reason, you can still fetch any URL generator from the registry, which Anton suggested.

Reply:
> good point, but introducing api like this we would still be introducing cyclic dependencies just in non-explicit way.

Consider an example with `Discover <-> Dashboard`
```js
// Dashboard wants to create a link to Dashboard
share.url.shortUrls.create({
  slug: 'my-discover-query', // Optional human-readable URL slug.
  generator: 'DISCOVER_DEEP_LINKS',
  state: {
    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx',
  },
});
```
How can we be sure that the ` generator id` and `state` are valid?
It can be changed at any time by Discover, but Dashboard doesn't know about it. TypeScript won't help here: Basel doesn't allow circular imports between packages.

> Discover <-> Dashboard

Maybe it can wait for the Domain-based folder structure ?https://github.com/elastic/kibana/issues/71566#issuecomment-688887617
In this case, Discover and Dashboard plugins can share `folders` within [Analyse domain](https://github.com/elastic/kibana/issues/71566#issue-656209142)

Reply:
> TypeScript won't help here: Basel doesn't allow circular imports between packages.

Should we consider runtime-based validation for this then? 
Btw, @streamich, I think this is needed anyway because we want to expose URL generation as an API, right? So should we look into configuring state shape like `@kbn/config-schema` does?

Reply:
the right solution is probably not a service like this but smaller plugins.

Reply:
I'm a bit confused, what is the outcome of this thread? How do you suggest to create a short URL?

This proposed way is due to us wanting to combine URL generators with short URL service, so that short URL service can do migrations automatically.

Reply:
TypeScript types can be imported from some third plugin/package.

> ... because we want to expose URL generation as an API, right?

Chatting with Peter today, it does not look like we want to expose URL generatrs as an API. Instead there will be a client-side redirect endpoint:

```
/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x
```